comparator/src/main/java/org/mobilitydata/gtfsvalidator/outputcomparator/util/ValidationReport.java

```diff
@@ -0,0 +1,120 @@
+ * limitations under the License.
+ */
+
+package org.mobilitydata.gtfsvalidator.outputcomparator.util;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/** Used to deserialize a validation report */
+public class ValidationReport {
+  private final List<NoticeAggregate> notices;
+
+  private ValidationReport(List<NoticeAggregate> notices) {
+    this.notices = notices;
+  }
+
+  /**
+   * Exports the integration test report (map of String, Object) as json.
+   *
+   * @param integrationReportData integration report content
+   * @param outputBase base path to output
+   * @param integrationReportName integration report name
+   * @throws IOException if an I/O error occurs writing to or creating the file
+   */
+  public static void exportIntegrationReportAsJson(
+      ImmutableMap<String, Object> integrationReportData,
+      String outputBase,
+      String integrationReportName)
+      throws IOException {
+    Gson gson = new GsonBuilder().serializeNulls().create();
+    Files.write(
+        Paths.get(outputBase, integrationReportName),
+        gson.toJson(integrationReportData).getBytes(StandardCharsets.UTF_8));
+  }
+
+  public List<NoticeAggregate> getNotices() {
+    return notices;
+  }
+
+  /**
+   * Return the set of error codes in a validation report
+   *
+   * @return the set of error codes in a validation report
+   */
+  public Set<String> getErrorCodes() {
+    Set<String> errorCodes = new HashSet<>();
+    for (NoticeAggregate noticeAggregate : getNotices()) {
+      if (noticeAggregate.isError()) {
+        errorCodes.add(noticeAggregate.getCode());
+      }
+    }
+    return errorCodes;
+  }
+
+  /**
+   * Compares two validation reports: returns true if they contain the same set of error codes
+   *
+   * @param otherValidationReport the other {@code ValidationReport}
+   * @return returns true if thw two {@code ValidationReport} contain the same set of error codes,
+   *     false otherwise
+   */
+  public boolean hasSameErrorCodes(ValidationReport otherValidationReport) {
+    return getErrorCodes().equals(otherValidationReport.getErrorCodes());
+  }
+
+  /**
+   * Compares two validation reports: returns the number of new error codes introduces in the
+   * validation report provided as parameter.
+   *
+   * @param otherValidationReport the other {@code ValidationReport}
+   * @return returns the number of new error codes introduces in the * validation report provided as
+   *     parameter.
+   */
+  public int getNewErrorCount(ValidationReport otherValidationReport) {
+    return (int)
+        otherValidationReport.getErrorCodes().stream()
+            .filter(errorCode -> !getErrorCodes().contains(errorCode))
+            .count();
+  }
```

Comment:
Try to measure the time complexity of this function and think how you can improve it.

Reply:
1. It seems that for loop are faster than sequential streams (https://jaxenter.com/java-performance-tutorial-how-fast-are-the-java-8-streams-118830.html). 

2. `.contains` has a complexity O(n) on an ArrayList - hence a O(n2) compleixty for the submitted function (in the worst case if the two lists contain `n` elements).

The following sample could possibly reduce the complexity of this function to O(n), provided `getErrorCodesgetErrorCodes()` returns `Map(String, Integer)`.

```java
  public int getNewErrorCount(ValidationReport otherValidationReport) {
    int newErrorCount = 0;
    for (String errorCode : getErrorCodes().keySet()) { // O(n)
      if (otherValidationReport.getErrorCodes().get(errorCode) == null) { // O(1)
        newErrorCount +=1;
      }
    }
    return newErrorCount;
  }
```

Please let me know if my reasoning wrong.

Reply:
Those lines have much higher complexity. Could you guess why?


```
for (String errorCode : otherValidationReport.getErrorCodes().keySet()) { // amortized: O(n), worst: O(n^2)
      if (getErrorCodes().get(errorCode) == null) { // amortized: O(n), worst: O(n^2)
        newErrorCount += 1;
      }
    }
```

It also will not compile because `getErrorCodes()` returns a `Set` that does not have `get()` and `keySet()`.

I see two possible solutions here.

1. Use a sorted set (TreeSet) that will give you a guaranteed O(n log n).
2. Use a HashSet but rewrite the code to make it amortized O(n). The worst case of O(n^2) is not realistic.

Reply:
> Use a sorted set (TreeSet) that will give you a guaranteed O(n log n).


Thanks! I provided changes in https://github.com/MobilityData/gtfs-validator/pull/848/commits/571ab245fe43d7267d7161e8e4270709aa163378

Reply:
But you have not changed getNewErrorCount and you still have O(n * n).